dbms/src/Debug/dbgKVStore/dbgFuncInvestigator.cpp

```diff
@@ -0,0 +1,289 @@
+
+#include <unordered_map>
+#include <unordered_set>
+
+namespace DB
+{
+
+struct MatchResult
+{
+    String mapped_database_name;
+    String table_name;
+    std::vector<std::pair<UInt64, UInt64>> in_default;
+    std::vector<std::pair<UInt64, UInt64>> in_write;
+    std::vector<UInt64> in_lock;
+    std::unordered_map<UInt64, RegionPtr> regions;
+
+    std::string toString() const
+    {
+        FmtBuffer fmt_buf;
+        FmtBuffer fmt_buf1;
+        FmtBuffer fmt_buf2;
+        FmtBuffer fmt_buf3;
+        for (const auto & a : in_default)
+        {
+            fmt_buf1.fmtAppend("{}-{}:", a.first, a.second);
+        }
+        for (const auto & a : in_write)
+        {
+            fmt_buf2.fmtAppend("{}-{}:", a.first, a.second);
+        }
+        for (const auto & a : in_lock)
+        {
+            fmt_buf3.fmtAppend("{}:", a);
+        }
+        fmt_buf.fmtAppend(
+            "default_cf {}; write_cf {}; lock_cf {}; ",
+            fmt_buf1.toString(),
+            fmt_buf2.toString(),
+            fmt_buf3.toString());
+        for (const auto & [region_id, region] : regions)
+        {
+            fmt_buf.fmtAppend(
+                "region {} {}, tikv_range: {}; ",
+                region_id,
+                region->getDebugString(),
+                region->getRange()->toDebugString());
+        }
+        return fmt_buf.toString();
+    }
+};
+
+// tablePrefix_rowPrefix_tableID_rowID
+static void findKeyInKVStore(Context & context, const ASTs & args, DBGInvoker::Printer output)
+{
+    MatchResult result;
+    const String & database_name_raw = typeid_cast<const ASTIdentifier &>(*args[0]).name;
+    const String & table_name = typeid_cast<const ASTIdentifier &>(*args[1]).name;
+
+    auto maybe_database_name = mappedDatabaseWithOptional(context, database_name_raw);
+    if (maybe_database_name == std::nullopt)
+    {
+        output(fmt::format("Database {} not found.", database_name_raw));
+        return;
+    }
+    result.mapped_database_name = maybe_database_name.value();
+    auto mapped_database_name = result.mapped_database_name;
+    result.table_name = table_name;
+
+    auto & tmt = context.getTMTContext();
+    auto & kvstore = *tmt.getKVStore();
+    auto schema_syncer = tmt.getSchemaSyncerManager();
+    auto storage = tmt.getStorages().getByName(mapped_database_name, table_name, false);
+    if (storage == nullptr)
+    {
+        output(fmt::format("can't find table {} {}", mapped_database_name, table_name));
+        return;
+    }
+
+    auto table_info = storage->getTableInfo();
+    schema_syncer->syncTableSchema(context, table_info.keyspace_id, table_info.id);
+    if (table_info.partition.num > 0)
+    {
+        for (const auto & def : table_info.partition.definitions)
+        {
+            schema_syncer->syncTableSchema(context, table_info.keyspace_id, def.id);
+        }
+    }
+
+    auto table_id = table_info.id;
+
+    TiKVKey start_key, end_key;
+    HandleID start_handle, end_handle;
```

Comment:
Define variables just before they are going to be used.

Reply:
fixed